### ISSUE TITLE
Don't use root level filter for search filters

### ISSUE THREAD

**Issue by pickypg** [ElasticLINQ/ElasticLINQ#71]
The current generation of queries that supply a `Filter` will result in a root level `filter`. For example, if you had:

``` cs
query.Query(q => q.Number > 10).Where(q => q.State == "WA");
```

then this generates something along the lines of

``` json
{
  "query" : {
    "range" : {
      "Number" : {
        "gt" : 10
      }
    }
  },
  "filter" : {
    "term" : {
      "State" : "WA"
    }
  }
}
```

Note: it's not really useful to score a `range` query as it either matches or not, but that's not the point.

However, using the root-level `filter` is actually using something called [Post Filter](https://www.elastic.co/guide/en/elasticsearch/guide/current/_post_filter.html), which is a post processing filter. This is useful when you want to pass certain hits to facets/aggregations, but then you need to prune them from being returned within the overall response (a rare need). What you actually want is a [`filtered` query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-filtered-query.html) that itself contains a `filter` and/or a `query`.

The above _should_ become:

``` json
{
  "query" : {
    "filtered" : {
      "query" : {
        "Number" : {
          "range" : {
            "gt" : 10
          }
        }
      },
      "filter" : {
        "term" : {
          "State" : "WA"
        }
      }
    }
  }
}
```

This will perform the `filter` before any query (or hits being returned), which is when you want it to happen.

**Comment by damieng** [comment]
I can see that filtered is useful if you're doing .Query and .Where (filter) together but I haven't been able to find any information as to why you'd want to add that when all you have is a filter. https://www.elastic.co/guide/en/elasticsearch/guide/current/_queries_and_filters.html says "As a general rule, use query clauses for full-text search or for any condition that should affect the relevance score, and use filter clauses for everything else."

I can't specifically find anything that says doing a top-level filter is the equivalent of the post_filter. The page linked above implies the opposite:

"Thanks to the inverted index, a simple query that matches just a few documents may perform as well or better than a cached filter that spans millions of documents. In general, however, a cached filter will outperform a query, and will do so consistently.

The goal of filters is to reduce the number of documents that have to be examined by the query."

**Comment by pickypg** [comment]
Those "filter" clauses are not the root level filter clause that it's suggesting here, rather they are the kind shown here. Specifically, from [Elasticsearch: The Definitive Guide](https://www.elastic.co/guide/en/elasticsearch/guide/current/_post_filter.html) at the bottom under "Performance Consideration":

> Use a `post_filter` only if you need to differentially filter search results and aggregations. Sometimes people will use `post_filter` for regular searches.
> 
> Don’t do this! The nature of the `post_filter` means it runs after the query, so any performance benefit of filtering (such as caches) is lost completely.

In terms of where `filter` was renamed to `post_filter`, it was back in [Elasticsearch 1.0.0](https://www.elastic.co/guide/en/elasticsearch/reference/current/_search_requests.html):

> Also, the top-level `filter` parameter in search has been renamed to `post_filter`, to indicate that it should not be used as the primary way to filter search results (use a `filtered` query instead), but only to filter results AFTER facets/aggregations have been calculated.

**Comment by damieng** [comment]
The Elastic page I linked to could do with some notes to that effect.

**Comment by pickypg** [comment]
Somewhat agree. Fortunately this problem will fix itself in ES 2.0 as the root level `filter` will be completely removed, with only `post_filter` remaining.

**Comment by damieng** [comment]
Fully checked in and adjusted the existing unit tests.  Will check to see if it needs any additional coverage when I'm in the office tomorrow.

Thanks!
